Android standalone app crashing on Android X

Hi expo team, I did update SDK yesterday. (35 to 36). Android build does not work on android x versions. The application gives error after splash screen. "Application keep stopping". and no error. I'm just getting unsafe lifecycle methods warning from GridList and Hyperlink packages.

I uninstalled and tried these packages, but the problem persists. But there is a detail; The application opens smoothly on first installation. But if you close it through the task manager it won't open again.

Expo CLI 3.11.1 environment info:
    System:
      OS: macOS 10.15.1
      Shell: 5.7.1 - /bin/zsh
    Binaries:
      Node: 10.17.0 - /usr/local/opt/node@10/bin/node
      Yarn: 1.19.1 - /usr/local/bin/yarn
      npm: 6.11.3 - /usr/local/opt/node@10/bin/npm
      Watchman: 4.9.0 - /usr/local/bin/watchman
    IDEs:
      Android Studio: 3.5 AI-191.8026.42.35.5900203
      Xcode: 11.1/11A1027 - /usr/bin/xcodebuild
    npmPackages:
      expo: ^36.0.0 => 36.0.0 
      react: 16.9.0 => 16.9.0 
      react-native: https://github.com/expo/react-native/archive/sdk-36.0.0.tar.gz => 0.61.4 
      react-navigation: ^4.0.10 => 4.0.10 
    npmGlobalPackages:
      expo-cli: 3.11.1

a lot of my users are facing the same issue on Android/standalone.
Finally got the crash myself (S10e, Android 9).
Logcat points to react-native-svg (I am using v9.13.3) - maybe this is helpful for the Expo devs:

01-08 17:40:03.482 27765 27895 E unknown:ViewManager: Error while updating prop width
01-08 17:40:03.482 27765 27895 E unknown:ViewManager: java.lang.reflect.InvocationTargetException
01-08 17:40:03.482 27765 27895 E unknown:ViewManager: 	at java.lang.reflect.Method.invoke(Native Method)
01-08 17:40:03.482 27765 27895 E unknown:ViewManager: 	at com.facebook.react.uimanager.ViewManagersPropertyCache$PropSetter.updateShadowNodeProp(ViewManagersPropertyCache.java:3)
01-08 17:40:03.482 27765 27895 E unknown:ViewManager: 	at com.facebook.react.uimanager.ViewManagerPropertyUpdater$FallbackShadowNodeSetter.setProperty(ViewManagerPropertyUpdater.java:2)
01-08 17:40:03.482 27765 27895 E unknown:ViewManager: 	at com.facebook.react.uimanager.ViewManagerPropertyUpdater.updateProps(ViewManagerPropertyUpdater.java:14)
01-08 17:40:03.482 27765 27895 E unknown:ViewManager: 	at com.facebook.react.uimanager.ReactShadowNodeImpl.updateProperties(ReactShadowNodeImpl.java:1)
01-08 17:40:03.482 27765 27895 E unknown:ViewManager: 	at com.facebook.react.uimanager.UIImplementation.createView(UIImplementation.java:11)
01-08 17:40:03.482 27765 27895 E unknown:ViewManager: 	at com.facebook.react.uimanager.UIManagerModule.createView(UIManagerModule.java:5)
01-08 17:40:03.482 27765 27895 E unknown:ViewManager: 	at java.lang.reflect.Method.invoke(Native Method)
01-08 17:40:03.482 27765 27895 E unknown:ViewManager: 	at com.facebook.react.bridge.JavaMethodWrapper.invoke(JavaMethodWrapper.java:18)
01-08 17:40:03.482 27765 27895 E unknown:ViewManager: 	at com.facebook.react.bridge.JavaModuleWrapper.invoke(JavaModuleWrapper.java:2)
01-08 17:40:03.482 27765 27895 E unknown:ViewManager: 	at com.facebook.react.bridge.queue.NativeRunnable.run(Native Method)
01-08 17:40:03.482 27765 27895 E unknown:ViewManager: 	at android.os.Handler.handleCallback(Handler.java:873)
01-08 17:40:03.482 27765 27895 E unknown:ViewManager: 	at android.os.Handler.dispatchMessage(Handler.java:99)
01-08 17:40:03.482 27765 27895 E unknown:ViewManager: 	at com.facebook.react.bridge.queue.MessageQueueThreadHandler.dispatchMessage(MessageQueueThreadHandler.java:1)
01-08 17:40:03.482 27765 27895 E unknown:ViewManager: 	at android.os.Looper.loop(Looper.java:214)
01-08 17:40:03.482 27765 27895 E unknown:ViewManager: 	at com.facebook.react.bridge.queue.MessageQueueThreadImpl$4.run(MessageQueueThreadImpl.java:8)
01-08 17:40:03.482 27765 27895 E unknown:ViewManager: 	at java.lang.Thread.run(Thread.java:764)
01-08 17:40:03.482 27765 27895 E unknown:ViewManager: Caused by: java.lang.NullPointerException: Attempt to invoke interface method 'boolean com.facebook.react.bridge.Dynamic.isNull()' on a null object reference
01-08 17:40:03.482 27765 27895 E unknown:ViewManager: 	at com.facebook.react.uimanager.LayoutShadowNode$MutableYogaValue.setFromDynamic(LayoutShadowNode.java:1)
01-08 17:40:03.482 27765 27895 E unknown:ViewManager: 	at com.facebook.react.uimanager.LayoutShadowNode.setWidth(LayoutShadowNode.java:2)
01-08 17:40:03.482 27765 27895 E unknown:ViewManager: 	... 17 more

01-08 17:40:03.507 27765 27765 D AndroidRuntime: Shutting down VM
01-08 17:40:03.508 27765 27765 E AndroidRuntime: FATAL EXCEPTION: main
01-08 17:40:03.508 27765 27765 E AndroidRuntime: Process: at.lszb.app, PID: 27765
01-08 17:40:03.508 27765 27765 E AndroidRuntime: java.lang.RuntimeException: Expo encountered a fatal error: Error while updating property 'width' in shadow node of type: RNSVGSvgView
01-08 17:40:03.508 27765 27765 E AndroidRuntime: 	at host.exp.exponent.experience.a$b.run(BaseExperienceActivity.java:7)
01-08 17:40:03.508 27765 27765 E AndroidRuntime: 	at android.os.Handler.handleCallback(Handler.java:873)
01-08 17:40:03.508 27765 27765 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:99)
01-08 17:40:03.508 27765 27765 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:214)
01-08 17:40:03.508 27765 27765 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:7037)
01-08 17:40:03.508 27765 27765 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
01-08 17:40:03.508 27765 27765 E AndroidRuntime: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:494)
01-08 17:40:03.508 27765 27765 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:965)

@strohmann I *have* seen crashes because of bad OTA updates, but those caused errors like the following for me:

```
12-26 21:58:33.473 10444 10444 E AndroidRuntime: FATAL EXCEPTION: main
12-26 21:58:33.473 10444 10444 E AndroidRuntime: java.lang.RuntimeException: Expo encountered a fatal error: Attempting to call JS function on a bad application bundle: RCTDeviceEventEmitter.emit()
12-26 21:58:35.786 10651 10651 E AndroidRuntime: FATAL EXCEPTION: main
12-26 21:58:35.786 10651 10651 E AndroidRuntime: java.lang.RuntimeException: Unable to resume activity {com.example.myapp/host.exp.exponent.MainActivity}: java.lang.RuntimeException: Expo encountered a fatal error: Attempting to call JS function on a bad application bundle: AppRegistry.runApplication()
12-26 21:58:35.786 10651 10651 E AndroidRuntime: Caused by: java.lang.RuntimeException: Expo encountered a fatal error: Attempting to call JS function on a bad application bundle: AppRegistry.runApplication()
```

and

```
12-26 21:58:44.934 10907 10907 E AndroidRuntime: FATAL EXCEPTION: main
12-26 21:58:44.934 10907 10907 E AndroidRuntime: java.lang.RuntimeException: Unable to resume activity {com.example.myapp/host.exp.exponent.MainActivity}: java.lang.RuntimeException: Expo encountered a fatal error: com.facebook.react.devsupport.JSException: Unexpected end of script
12-26 21:58:44.934 10907 10907 E AndroidRuntime: Caused by: java.lang.RuntimeException: Expo encountered a fatal error: com.facebook.react.devsupport.JSException: Unexpected end of script
```

rather than:
```
Error while updating property 'fontSize' in shadow node of type: RCTText
```

--- a/android/expoview/src/main/java/host/exp/exponent/experience/ExperienceActivity.java
+++ b/android/expoview/src/main/java/host/exp/exponent/experience/ExperienceActivity.java
@@ -10,8 +10,6 @@ import android.content.Intent;
 import android.net.Uri;
 import android.os.Bundle;
 import android.os.Handler;
-import androidx.core.app.NotificationCompat;
-import androidx.core.content.ContextCompat;
 import android.text.TextUtils;
 import android.view.View;
 import android.view.animation.AccelerateInterpolator;
@@ -36,6 +34,8 @@ import java.util.List;
 import javax.annotation.Nullable;
 import javax.inject.Inject;
 
+import androidx.core.app.NotificationCompat;
+import androidx.core.content.ContextCompat;
 import de.greenrobot.event.EventBus;
 import host.exp.exponent.AppLoader;
 import host.exp.exponent.Constants;
@@ -99,6 +99,7 @@ public class ExperienceActivity extends BaseExperienceActivity implements Expone
   private boolean mIsShellApp;
   private String mIntentUri;
   private boolean mIsReadyForBundle;
+  private boolean mWillBeReloaded = false;
 
   private RemoteViews mNotificationRemoteViews;
   private Handler mNotificationAnimationHandler;
@@ -396,7 +397,7 @@ public class ExperienceActivity extends BaseExperienceActivity implements Expone
     Analytics.logEventWithManifestUrlSdkVersion(Analytics.LOAD_EXPERIENCE, mManifestUrl, mSDKVersion);
 
     ExperienceActivityUtils.updateOrientation(mManifest, this);
-    ExperienceActivityUtils.overrideUserInterfaceStyle(mManifest, this);
+    mWillBeReloaded = ExperienceActivityUtils.overrideUserInterfaceStyle(mManifest, this);
     addNotification(kernelOptions);
 
     ExponentNotification notificationObject = null;
@@ -472,7 +473,10 @@ public class ExperienceActivity extends BaseExperienceActivity implements Expone
     // by this point, setManifest should have also been called, so prevent
     // setLoadingScreenManifest from showing a rogue loading screen
     mShouldShowLoadingScreenWithOptimisticManifest = false;
-    if (!isDebugModeEnabled()) {
+
+    // To prevents starting application twice, we start react instance only if we know that the current activity won't be restarted.
+    // Restart of the activity could be triggered by dark mode change.
+    if (!isDebugModeEnabled() && !mWillBeReloaded) {
       final boolean finalIsReadyForBundle = mIsReadyForBundle;
       AsyncCondition.wait(READY_FOR_BUNDLE, new AsyncCondition.AsyncConditionListener() {
         @Override

--- a/android/expoview/src/main/java/host/exp/exponent/utils/ExperienceActivityUtils.java
+++ b/android/expoview/src/main/java/host/exp/exponent/utils/ExperienceActivityUtils.java
@@ -5,6 +5,7 @@ package host.exp.exponent.utils;
 import android.app.Activity;
 import android.app.ActivityManager;
 import android.content.pm.ActivityInfo;
+import android.content.res.Configuration;
 import android.graphics.Bitmap;
 import android.graphics.Color;
 import android.os.Build;
@@ -21,6 +22,7 @@ import androidx.core.view.ViewCompat;
 
 import host.exp.exponent.ExponentManifest;
 import host.exp.exponent.analytics.EXL;
+import host.exp.expoview.R;
 
 import org.json.JSONException;
 import org.json.JSONObject;
@@ -56,10 +58,22 @@ public class ExperienceActivityUtils {
 
   // region user interface style - light/dark/automatic mode
 
-  public static void overrideUserInterfaceStyle(JSONObject manifest, AppCompatActivity activity) {
+  /**
+   * Returns true if activity will be reloaded after night mode change.
+   * Otherwise returns false.
+   **/
+  public static boolean overrideUserInterfaceStyle(JSONObject manifest, AppCompatActivity activity) {
     String userInterfaceStyle = readUserInterfaceStyleFromManifest(manifest);
     int mode = nightModeFromString(userInterfaceStyle);
+    boolean isNightModeCurrentlyOn = activity.getResources().getBoolean(R.bool.dark_mode);
+    boolean willBeReloaded = false;
+    if (mode != AppCompatDelegate.MODE_NIGHT_AUTO) {
+      willBeReloaded = isNightModeCurrentlyOn && mode == AppCompatDelegate.MODE_NIGHT_NO
+        || !isNightModeCurrentlyOn && mode == AppCompatDelegate.MODE_NIGHT_YES;
+    }
+
     activity.getDelegate().setLocalNightMode(mode);
+    return willBeReloaded;
   }
 
   private static int nightModeFromString(@Nullable String userInterfaceStyle) {

--- a/android/expoview/src/main/res/values-night/dark_mode.xml
+++ b/android/expoview/src/main/res/values-night/dark_mode.xml
@@ -1,0 +1,3 @@
+<resources>
+  <bool name="dark_mode">true</bool>
+</resources>

--- a/android/expoview/src/main/res/values/dark_mode.xml
+++ b/android/expoview/src/main/res/values/dark_mode.xml
@@ -1,0 +1,3 @@
+<resources>
+  <bool name="dark_mode">false</bool>
+</resources>

